src/grain/payouts.js

```diff
@@ -0,0 +1,180 @@
+ * to serious precision issues in practice.
+ */
+export const _multiply = (x: Grain, p: number) => {
+  const floated = Number(x) * p;
+  // $ExpectFlowError
+  return BigInt(Math.floor(floated));
+};
+
+/**
+ * Distribute an amount of Grain, strictly in proportion to the scores.
+ *
+ * This might be a good fit for distributing a fixed grain per week directly
+ * in proportion to the weekly cred score.
+ */
+export function fixedAmount(
+  harvestAmount: Grain,
+  scores: Map<NodeAddressT, number>
+): Map<NodeAddressT, Grain> {
+  if (harvestAmount < ZERO) {
+    throw new Error(`invalid harvestAmount: ${String(harvestAmount)}`);
+  }
+  if (scores.size === 0) {
+    return new Map();
+  }
+  const totalScore = sum(Array.from(scores.values()));
+  if (totalScore === 0) {
+    const perCapita = _multiply(harvestAmount, 1 / scores.size);
+    return MapUtil.mapValues(scores, () => perCapita);
+  }
+  const computeShare = (_, score) =>
+    _multiply(harvestAmount, score / totalScore);
+  return MapUtil.mapValues(scores, computeShare);
+}
+
+/**
+ * Distribute Grain so that each contributor's lifetime earnings
+ * are at least a fixed multiple of their score.
+ *
+ * Parameters:
+ *
+ * grainPerScore: Target amount of grain for each unit of score. This is
+ *   recorded as a number, and ignores grain precision (e.g. if
+ *   grainPerScore===2, then we are aiming to distribute two "full" grain per
+ *   unit of score, which might be tracked as 2n * 10n**18n).
+ *
+ * scores: The map from users' addresses to their scores. This could be
+ *   lifetime cred scores, it could be this week's scores, etc.
+ *
+ * earnings: The map from users' addresses to their lifetime distributed payout.
+ *   This information is used to determine which users have been "underpaid"
+ *   relative to the payout target, and by how much.
+ *
+ * Since this strategy provides a consistent reward per unit cred, it means that
+ * as the activity level (and thus, cred generation) in a project increases, the
+ * rewards also increase. This avoids the problem with fixed-reward strategies,
+ * where more participants mean less reward per person.
+ *
+ * If some users have been historically over-paid (e.g. because the
+ * target grainPerScore was higher in the past, or a weight change decreased
+ * their score), they might go for a long time without any payouts.
+ *
+ * Also, if users find a way to game/inflate cred scores, using this payout
+ * strategy will cause grain to inflate as well.
+ */
+export function fixedRatio(
+  grainPerScore: number,
+  scores: Map<NodeAddressT, number>,
+  earnings: Map<NodeAddressT, Grain>
+): Map<NodeAddressT, Grain> {
+  if (!isFinite(grainPerScore) || grainPerScore < 0) {
+    throw new Error(`invalid grainPerScore: ${grainPerScore}`);
+  }
+  return MapUtil.mapValues(scores, (addr, score) => {
+    const earned = earnings.get(addr) || ZERO;
+    const target = _multiply(ONE, grainPerScore * score);
```

Comment:
When I first read "multiply" it makes it seem like thats the function actually doing the multiplication to calculate the target grain, when in reality the actual calculation is just the `grainPerScore * score`. 

Seems like the intent of this is to "convert" the float value to the BigInt value for grain. Would make it more readable if you wrapped this in a helper function, something like:
```ts
const toGrain = (score: number): Grain => _multiply(ONE, score);
```

I imagine it would get re-used in other places as well. Seems like you already had something like this in the tests

Reply:
Hmm, the "conversion" is actually a multiplication though. Since `ONE === 10n**18n`. If we ignore the type conversions, the line is something like `10n**18n * grainPerScore * score`. That said, I can see how a `floatToGrain` type method could be useful; you're right that I define such a method in the test code. I'll sleep on it and maybe add it in the morning.

Also, generally speaking, we use "request changes" when the reviewer wants to block approval pending fixing a serious issue; here's a good example of [@wchargin blocking a PR that introduced a security flaw](https://github.com/sourcecred/sourcecred/pull/1265#pullrequestreview-273471576). When there's just nits, we'll usually either just comment, or [nit-and-approve](https://github.com/sourcecred/docs/blob/master/culture/review_culture.md#consider-nit-and-approve).

Reply:
@hammadj I factored your suggestion into another PR: https://github.com/sourcecred/sourcecred/pull/1715